packages/compiler/src/render3/view/api.ts

```diff
@@ -85,11 +85,12 @@ export interface R3DirectiveMetadata {
     usesOnChanges: boolean;
   };
 
+  // TODO(required-inputs): add `required` to the `inputs` interface.
   /**
    * A mapping of inputs from class property names to binding property names, or to a tuple of
    * binding property name and class property name if the names are different.
    */
-  inputs: {[field: string]: string|[string, string]};
+  inputs: {[field: string]: {classPropertyName: string, bindingPropertyName: string}};
```

Comment:
I would suggest creating the interface:

```typescript
export interface R3InputMetadata {
  classPropertyName: string;
  bindingPropertyName: string;
}
```

Then the `inputs` declaration could become:
```suggestion
  inputs: R3InputMetadata[],
```

Using an array instead of an object also makes the downstream iteration easier, and avoids the need to do `Object.key` operations to iterate it.